packages/actions-spec/README.md

```diff
@@ -441,11 +459,21 @@ A `POST` response with an HTTP `OK` JSON response should include a body payload
 of:
 
 ```ts filename="ActionPostResponse"
-export interface ActionPostResponse {
+/**
+ * Response body payload returned from the Action POST Request
+ */
+export interface ActionPostResponse<T extends ActionType = ActionType> {
   /** base64 encoded serialized transaction */
   transaction: string;
   /** describes the nature of the transaction */
   message?: string;
+  links?: {
+    /**
+     * the next action in a successive chain of actions to be obtained and/or rendered
+     * to the user after the previous was successful
+     */
+    next: string | NextAction<T | "action">;
```

Comment:
May be the `embedded` is not the best noun, but the idea should be clear

Reply:
We can avoid the string easily with the suggestion you made in [this comment](https://github.com/solana-developers/solana-actions/pull/18#discussion_r1697431298) :)

To me, making this union types accomplishes the same thing but with more complexity in the type and makes a developer have to write more verbose code to accomplish the same thing. Especially since we are only supporting two types of "next actions" right now.

If and when more types are supported, we might then need the more complicated type and verbosity you are suggesting here. It would be easily compatible with the current proposal AND your suggestion here by performing a union type based on the `type` field then.

Reply:
I think the comment you are referring in comment is actually related, but expresses the same concept, I'm proposing here. The idea of the comment was to enforce `type` attribute being mandatory in `NextAction`, since it was optional. This way action developers will be enforced to set the `type` and client can then deterministically distinguish if the next action is `completed` or `action`.

While the current approach accomplishes the task, we should also consider extensibility, type-safety and deterministic interpretation of data structures in the future.

### Main Problems with raw strings and primitive values in general

The primary issue with using raw strings and primitive values in general is extensibility. It's generally best practice to wrap plain values in objects to allow for additional attributes or have other union options that may be expressed with the same primitive type, but have different semantics.

####  Example scenarios illustrating the potential problems:

1. **Supporting GET URLs in the future:**
   - **Current approach:**
     ```typescript
     export interface ActionPostResponse<T extends ActionType = ActionType> {
       transaction: string;
       message?: string;
       links?: {
         next: string | NextAction<T | "action">;
       };
     }
     ```
   - If we decide to support GET URLs in the future, a plain string won't suffice, and we will need to add an object.
     ```typescript
     export interface GetNextActionLink {
       type: "get";
       href: string;
     }
     ```
   - This would require maintaining the raw string for backward compatibility and introducing a new object, leading to an inconsistent interface - POST will be raw string, GET will be an object.

2. **Adding extra parameters to POST:**
   - **Current Approach:**
     ```typescript
     export interface ActionPostResponse<T extends ActionType = ActionType> {
       transaction: string;
       message?: string;
       links?: {
         next: string | NextAction<T | "action">;
       };
     }
     ```
   - If we want to add extra parameters to a POST action, using a raw string will make this impossible. We would need to introduce an object to handle additional parameters.
     ```typescript
     export interface PostNextActionLink {
       type: "post";
       href: string;
       extraParams?: { [key: string]: any };
     }
     ```
   - Again, this would require keeping the raw string for backward compatibility and introducing a new object, resulting in inconsistency.

In both cases, maintaining backward compatibility means supporting the raw string while introducing new objects as proposed. This will complicate the interface and make it look inconsistent, decreasing the DX.

### Need for type attribute in union type members

When you have multiple objects, a clear way to distinguish the type client-side is necessary for type-safe switch-casing. Using `typeof` won't work effectively. Instead, a `type` attribute ensures deterministic interpretation and easy handling.

#### Example:

**Proposed approach:**
```typescript
export type NextActionLink<T extends ActionType> = PostNextActionLink | GetNextActionLink | EmbeddedNextAction<T>;

export interface PostNextActionLink {
  type: "post";
  href: string;
}

export interface GetNextActionLink {
  type: "get";
  href: string;
}

export interface EmbeddedNextAction<T extends ActionType> {
  type: "embedded";
  action: NextAction<T>;
}

function handleNextAction(next: NextActionLink<any>) {
  switch (next.type) {
    case 'post':
      // Type safe handle post action with next.href
      break;
    case 'get':
      // Type safe handle get action with next.href
      break;
    case 'embedded':
      // Type safe handle embedded action with next.action
      break;
    default:
      // Handle unexpected case
      break;
  }
}

```

### References to data structures designed for similar problem space
Many well-designed APIs follow these best practices:

- **HAL:** Check how `next` is defined in HAL:
  [HAL Specification](https://stateless.group/hal_specification.html)
- **Farcaster Actions:** Look at how actions are defined in Farcaster, even for a single action type:
  [Farcaster Actions Specification](https://docs.farcaster.xyz/reference/actions/spec#properties)

These references are designed with the same reasons highlighted above: extensibility, consistency, and deterministic client-side interpretation during the future updates.

Reply:
I totally get understand your goal and desired implementation, I just think it is more complicated and verbose. Both of our recommendations accomplish the same thing and effectively provide the same type safety and DX (for both action developer and blink clients). 

As a dev building actions, I would prefer using my recommended implementation but I can also see why as a dev building blink clients, you would prefer yours. The accomplish the same goal in the end.

I'm fine with moving forward with yours so we can finalize the spec and get it shipped. Can you just suggest the changes on my PR?

Reply:
Merged your PR into mine :)